Mathlib/Combinatorics/Flow/Basic.lean

```diff
@@ -0,0 +1,232 @@
+/-
+Copyright (c) 2024 Niklas Mohrin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Niklas Mohrin
+-/
+import Mathlib.Algebra.BigOperators.Group.Finset
+import Mathlib.Combinatorics.Digraph.Basic
+
+/-!
+# Network Flows
+
+In this file we introduce network flows. A flow network can be seen as a directed graph where each
+arc of the graph has a fixed capacity. A flow from a source vertex `s` to a sink vertex `t` is an
+assignment of "flow units" to each edge, such that for all vertices other than the source and the
+sink, the incoming flow must equal the outgoing flow (_flow conservation_). The value of such an
+assignment is the amount of flow units leaving the source `s` and arriving at the sink `t`. In
+practice, we are usually interested in finding a flow of maximal value for a given network.
+
+## Main definitions
+
+- `Network`: A flow network assigning capacities to all pairs of vertices. Notably, capacities and
+             non-negative and the capacity from a vertex to itself is zero.
+- `PseudoFlow`, `PreFlow`, and `Flow`: An assignment of flow values for a given `Network`, abiding
+                                       increasingly strict constraints.
+
+## Notation
+
+For flows `f f' : N.PseudoFlow`, we define `f ≤ f'` as the _subflow relation_, indicating that for
+all `u v : V`, `f u v ≤ f' u v`.
+
+## References
+
+See [goldberg_network_1989] for a thorough introduction to the topic.
+
+## Tags
+
+flow, network, max-flow, maxflow, cut, min-cut, mincut, graph, subflow
+-/
+
+universe u_V u_R
+
+open BigOperators
+
+section NetworkDefinition
+
+variable (V : Type u_V) (R : Type u_R) [Zero R] [LE R]
+
+/-- A flow network assigning non-negative capacities to all pairs of vertices. -/
+structure Network extends Digraph V where
+  Adj _ _ := True
```

Comment:
what is the point of adding an Adj relation that makes the graph complete. Flow problems are more general

Reply:
This is not what this means. I am updating the default value for the `Adj` field of `Digraph` (see https://lean-lang.org/lean4/doc/struct.html#default-field-values). Please see my last two comments in this github thread for my thoughts on this

Reply:
Apologies, I forgot how lean allows new defaults for extended structures. I am happy to mark this comment resolved.